compiler/rustc_mir_transform/src/generator.rs

```diff
@@ -749,9 +749,15 @@ fn sanitize_witness<'tcx>(
         }
         let decl_ty = tcx.normalize_erasing_regions(param_env, decl.ty);
 
+        let is_uninhabited = tcx.is_ty_uninhabited_from(
+            tcx.parent_module(tcx.hir().local_def_id_to_hir_id(did.expect_local())).to_def_id(),
+            decl_ty,
+            param_env,
+        );
+
         // Sanity check that typeck knows about the type of locals which are
         // live across a suspension point
-        if !allowed.contains(&decl_ty) && !allowed_upvars.contains(&decl_ty) {
+        if !is_uninhabited && !allowed.contains(&decl_ty) && !allowed_upvars.contains(&decl_ty) {
```

Comment:
```suggestion
        if is_inhabited && !allowed.contains(&decl_ty) && !allowed_upvars.contains(&decl_ty) {
```

This if condition is sufficiently complex that, to me at least, it is easier to read this way.

Alternatively, I would consider:

```rust
if is_uninhabited || allowed.contains(&decl_ty) || allowed_upvars.contains(&decl_ty) {
    // This type which appears in the generator either...
    // - is uninhabited, in which case it can't actually be captured at runtime
    // - appears in the approximation from thje static type (`allowed`)
    // - appears in the lsit of upvars ...
} else {
    span_bug!(...);
}
```

(only those comments are wrong)

Reply:
That said, maybe we should move the uninhabited check somewhere else so that we just *don't capture the type at the MIR level*?

Reply:
I looked at that a little at first and gave up. The right place to do this is probably in either `locals_live_across_suspend_points` or maybe in the `MaybeLiveLocals` MIR pass that `locals_live_across_suspend_points` uses. In `MaybeLiveLocals`, I didn't see a great way to get access to the `tcx` to ask if a type is uninhabited, but I wouldn't be surprised if there's an obvious way I don't know about.

Reply:
I like your rephrasing of the conditional, so I appliked your suggestion.

Reply:
#93313 exposes information about uninhabited types in call expressions to analyses running on MIR, and would be an another way to close the precision gap.

Reply:
@tmiasko - do you know if you change fixes this same issue as well? If so, I think yours is the more principled fix.

Reply:
It does fix the test case included here.

I understood the difference to be caused by modelling of calls with uninhabited calls as not returning in `handle_uninhabited_return`, so #93313 should provide a MIR equivalent.

Reply:
Awesome, thanks! I'll leave my version here for now, but if yours merges before this one I'll pull it out and rely on yours since it's better.